src/main/java/io/cdap/plugin/gcp/publisher/source/PubSubReceiver.java

```diff
@@ -0,0 +1,546 @@
+    } catch (IOException e) {
+      LOG.error("Unable to get credentials.");
+      stop("Unable to get credentials for receiver.", e);
+    }
+
+    //Create subscription if the topic is specified.
+    if (topic != null) {
+      createSubscription();
+    }
+
+    //Schedule tasks to set the message rate (Token Bucket algorithm) and start the receiver worker.
+    this.executor.scheduleAtFixedRate(this::updateMessageRateAndFillBucket, 0, 1, TimeUnit.SECONDS);
+    this.executor.submit(this::scheduleFetch);
+
+    LOG.info("Receiver started execution");
+  }
+
+  @Override
+  public void onStop() {
+    //Shutdown thread pool executor
+    if (executor != null && !executor.isShutdown()) {
+      executor.shutdown();
+      try {
+        executor.awaitTermination(30, TimeUnit.SECONDS);
+      } catch (InterruptedException e) {
+        LOG.error("InterruptedException while waiting for executor to shutdown.");
+      }
+    }
+
+    //Clean up subscriber stub used by the Google Cloud client.
+    if (subscriber != null && !subscriber.isShutdown()) {
+      subscriber.shutdown();
+      try {
+        subscriber.awaitTermination(30, TimeUnit.SECONDS);
+      } catch (InterruptedException e) {
+        LOG.error("InterruptedException while waiting for subscriber to shutdown.");
+      }
+    }
+
+    LOG.info("Receiver completed execution");
+  }
+
+  /**
+   * Create a new subscription (if needed) for the supplied topic.
+   *
+   * @throws IllegalArgumentException when the specified Topic does not exists
+   * @throws RuntimeException         when the SubscriptionAdminClient cannot be created.
+   * @throws ApiException             when a non-retryable exception is thrown by the Pub/Sub client
+   */
+  protected void createSubscription() {
+    int backoff = backoffConfig.getInitialBackoffMs();
+    int attempts = 5;
+
+    ApiException lastApiException = null;
+
+    while (!isStopped() && attempts-- > 0) {
+
+      try (SubscriptionAdminClient subscriptionAdminClient = getSubscriptionAdminClient()) {
+
+        int ackDeadline = 10; //10 seconds before resending the message.
+        subscriptionAdminClient.createSubscription(
+          subscription, topic, PushConfig.getDefaultInstance(), ackDeadline);
+        return;
+
+      } catch (ApiException ae) {
+
+        lastApiException = ae;
+
+        //If the subscription already exists, ignore the error.
+        if (ae.getStatusCode().getCode().equals(StatusCode.Code.ALREADY_EXISTS)) {
+          return;
+        }
+
+        //This error is thrown is the Topic does not exist.
+        // Call the stop method so the pipeline fails.
+        if (ae.getStatusCode().getCode().equals(StatusCode.Code.NOT_FOUND)) {
+          String message = String.format(MISSING_TOPIC_ERROR_MSG, topic, project);
+          stop(message, ae);
+          return;
+        }
+
+        //Retry if the exception is retriable.
+        if (ae.isRetryable()) {
```

Comment:
should this also be checking those retryable status codes from the top of the class?

Reply:
Hmm, good catch. I will use the same method call I used in the `fetchAndAck` method.